### PR TITLE
Feature/export selected

### PR DESCRIPTION
**Pull request by lloydmeta**
Addresses #163 

Works by:

- Adding an export-selected anchor selector parameter to the datatable template
- Adding a listener for clicked events on the table and updating the clicked row's style to be `selected`(natively supported by Jquery datatables)
- Adding a select all button
- Find the list of selected rows, get their respective ids, and then update the URL on the aforementioned anchor, disabling it if there are no items selected for export.

screen shot 2015-08-06 at 8 52 26 pm

### CODE REVIEW

**Comment by mauhiz** [comment]
The are a LOT of unrelated changes because a formatter went crazy

**Comment by xevix** [comment]
Awesome, this will help a lot in exporting several things quickly.

The flow I had in mind was something like:

1. Filter using the Search box on the right for just the sets of endpoints desired
2. Click some "select all" kind of button, and adjust selection if desired
3. Click export

The style changes themselves aren't an issue, but it's not possible to isolate the semantic changes. Maybe have 1 commit with the semantic changes, and a different one for the style if desired.

**Comment by lloydmeta** [comment]
@xevix cool, added a select all button.

**Comment by lloydmeta** [comment]
> The are a LOT of unrelated changes because a formatter went crazy

:( this is getting fairly old now.

**Comment by lloydmeta** [comment]
Reverted most of the style-changes caused by Scalariform.

2 are kept because they actually make sense (aligning the routes file and the keyword arguments of a method call)

**Comment by lloydmeta** [comment]
Updated screenshot with the "Select all" button
screen shot 2015-08-07 at 10 09 07

File: app/views/part/list.scala.html [hunk 1]
```diff
@@ -12,11 +12,63 @@
...
+        var toggleDownloadLink = function(enabled) {
+          if(enabled) {
+            exportSelectedLink.removeClass("disabled");
+            exportSelectedLink.removeAttr("disabled");
+          } else {
+            exportSelectedLink.attr("disabled", "disabled");
+          }
+        }
+
+	    exportSelectedLink.attr("disabled", "disabled");
```
Comment (xevix) [inline app/views/part/list.scala.html]: This initialization should probably be higher up with the var declarations for visibility.

**Comment by xevix** [comment]
Not a dealbreaker, but often a few quick clicks on selecting rows can lead to a text-area selection as seen in the URL field here. I like the idea of clicking directly on a row to select, but perhaps a selection column on the left, with or without checkboxes, might work better here. A user trying to select the URL for copy+paste will also find themselves inadvertently selecting a part.



The rest looks awesome. Great addition.

File: app/views/part/list.scala.html (hunk 1 above)
Comment (lloydmeta) [inline app/views/part/list.scala.html]: Tried it out, but didn't like how it mixed data declarations with side-effecting. Added a comment instead for clarity.

**Comment by lloydmeta** [comment]
> Not a dealbreaker, but often a few quick clicks on selecting rows can lead to a text-area selection as seen in the URL field here. I like the idea of clicking directly on a row to select, but perhaps a selection column on the left, with or without checkboxes, might work better here. A user trying to select the URL for copy+paste will also find themselves inadvertently selecting a part

Reimplemented with a checkbox column.

File: conf/messages.en [hunk 2]
```diff
@@ -71,9 +71,9 @@ parts.cacheGroup.none=None yet
 parts.section.alertMail=Alert mail settings
 parts.alertMail.condition=Condition
 parts.alertMail.condition.label=Trigger condition
-parts.alertMail.condition.absolute={0} 
+parts.alertMail.condition.absolute={0}
 parts.alertMail.condition.relative={0}%
-parts.alertMail.condition.or= or
```
Comment (mauhiz) [inline conf/messages.en]: I think this space actually matters

**Comment by mauhiz** [comment]
Very nice feature! I'll test it some more locally. Meanwhile, could you please reorganize your commits between formatting changes and the core of this PR?

File: conf/messages.en (hunk 2 above)
Comment (lloydmeta) [inline conf/messages.en]: Nah, it actually doesn't. It's currently setting inside a div that ends up removing all white space anyway.

**Comment by mauhiz** [comment]
The checkbox column is sortable but likely does not need to be.

**Comment by lloydmeta** [comment]
> The checkbox column is sortable but likely does not need to be.

The column was actually not sortable but DataTables was showing it as sortable (annoyingly). This is now fixed.

**Comment by mauhiz** [comment]
LGTM

**Comment by xevix** [comment]
@lloydmeta Awesome. One final nitpick is the following sequence:

1. Click select all
2. Manually deselect selected boxes
3. Select all checkbox should be uncleared

**Comment by lloydmeta** [comment]
@xevix good idea. sounds like a good way to improve the UX. done.

**Comment by xevix** [comment]
@lloydmeta It seems like the last commits changed existing behavior for this flow:

1. Click select all
2. Unselect an item
3. Select all remains checked, so that you can click it after to deselect all

As a reference I'm using GMail's behavior. This flow did seem to make sense, as when a person clicks select all, then deselects an item, it's more likely they may later wish to deselect all, rather than click select all again. Thoughts?

**Comment by lloydmeta** [comment]
Yeah, I changed the SelectAll checkbox behaviour on purpose based on logic. The idea is that when you check it, it selects everything and is directly tied to that state: unchecking any checkboxes causes the SelectAll checkbox to change to something other than "all selected; since a checkbox is binary, it would be unchecked. LinkedIn's messages select all checkbox does this.

Gmail's select all isn't a checkbox, it's a triple (possibly more?) state custom thing with a dropbox containing other actions (let's call it an uber-select-check). When you check it and unselect some items, it changes into a "some selected" state that is distinct from it's "all selected" state. 

Short of having data from user-testing, in our select-all-then-unselect-some scenario, I think it would be hard to say whether a simple checkbox should remain checked based on what a user is more likely to do because it involves successive second guesses on user intent/ability/mistakes. In addition, StackExchange has [a](http://stackoverflow.com/questions/522352/most-usable-select-all-deselect-all-checkbox-format) [couple](http://ux.stackexchange.com/questions/1382/check-all-uncheck-all-checkboxes-with-regards-to-usability) threads where people say this is an inherently tricky thing to decide. I think the confusion stems from the fact that the checkbox is simultaneously used as a way of invoking an action and as an indicator of current state (are all things checked?).

After reading a few posts, I'm in favour of having separate Select All and Select None buttons; this keeps things simple and never incurs the cost of extra click on behalf of the user.

**Comment by xevix** [comment]
Yeah, I wasn't saying to emulate the GMail checkbox's other behaviors which we don't need, just those related to the select all/none functionality. I'm certain that the efficacy of that UI had user-testing backing it as that's Google's general policy, but it's true we don't have access to said data.

Go with whichever one you think makes the most sense.

**Comment by lloydmeta** [comment]
I've pushed an update that implements 2 select toggle buttons.

At the end of the day, the checkmark solution implemented either way would have only resulted in at most 1 extra click required in case we guessed wrong, so it's not a huge deal. I think we should go with this for now, try it out, and if we don't like the real estate that this solution takes up, switch it to the tri-state checkbox à la Google or back to LinkedIn (fairly trivial)

screen shot 2015-08-16 at 5 37 34 pm

**Comment by xevix** [comment]
@lloydmeta Awesome, LGTM, merging.

**Pull request merged**